Bluetooth: Mesh: Sensor Server and Client sample (#2513)
nrfconnect/sdk-nrf: samples/bluetooth/mesh/sensor_client/README.rst

.. _bluetooth_mesh_sensor_client:

> Bluetooth: Mesh Sensor Client

Reviewer: ```suggestion
Bluetooth: Mesh sensor client
```

###############################

> The Bluetooth Mesh Sensor Client sample demonstrates how to set up a basic Bluetooth Mesh :ref:`bt_mesh_sensor_cli_readme` model application that gets sensor data from one :ref:`bt_mesh_sensor_srv_readme` model.

Reviewer: ```suggestion
The Bluetooth Mesh sensor client sample demonstrates how to set up a basic Bluetooth Mesh :ref:`bt_mesh_sensor_cli_readme` model application that gets sensor data from one :ref:`bt_mesh_sensor_srv_readme` model.
```

.. note::

>     This sample must be paired with :ref:`bluetooth_mesh_sensor_server` to show any functionality. A Client has no sensor data and is dependent on a Server to provide it.

Reviewer: ```suggestion
    This sample must be paired with :ref:`bluetooth_mesh_sensor_server` to show any functionality. A client has no sensor data and is dependent on a server to provide it.
```

Four Mesh sensor types are used in this sample:

1. Present device operating temperature - Published by the Server according to the publishing period of the server.
#. Relative runtime in a device operating temperature range - Periodically requested by the Client.
#. Presence detected - Published when a button is pressed on the Server board.

> #. Time since presence detected - Periodically requested by the Client.

Author: Changed in README for both Server and Client

> 1. :cpp:var:`bt_mesh_sensor_present_dev_op_temp` - Published by the Server according to the publishing period of the Server.
> #. :cpp:var:`bt_mesh_sensor_rel_runtime_in_a_dev_op_temp_range` - Periodically requested by the Client.
> #. :cpp:var:`bt_mesh_sensor_presence_detected` - Published when a button is pressed on the Server board.
> #. :cpp:var:`bt_mesh_sensor_time_since_presence_detected` - Periodically requested by the Client and published by the Server according to the publishing period of the Server.

Reviewer: ```suggestion
1. :cpp:var:`bt_mesh_sensor_present_dev_op_temp` - Published by the server according to its publishing period
#. :cpp:var:`bt_mesh_sensor_rel_runtime_in_a_dev_op_temp_range` - Periodically requested by the client
#. :cpp:var:`bt_mesh_sensor_presence_detected` - Published when a button is pressed on the server board
#. :cpp:var:`bt_mesh_sensor_time_since_presence_detected` - Periodically requested by the client and published by the server according to its publishing period
```

Overview
********

This sample is split into two source files:

* A :file:`main.c` file to handle initialization.
* One additional file for handling Mesh models, :file:`model_handler.c`.

Provisioning
============

The provisioning is handled by the :ref:`bt_mesh_dk_prov`, and can be done with the `nRF Mesh mobile app`_. It is also possible to configure models with the app.

Models
======

> The following table shows the Mesh Sensor Client composition data for this sample:

Reviewer: ```suggestion
The following table shows the Mesh sensor client composition data for this sample:
```

   +---------------+
   |  Element 1    |
   +===============+
   | Config Server |
   +---------------+
   | Health Server |
   +---------------+
   | Sensor Client |
   +---------------+

The models are used for the following purposes:

* Config Server allows configurator devices to configure the node remotely.
* Health Server provides ``attention`` callbacks that are used during provisioning to call your attention to the device.
  These callbacks trigger blinking of the LEDs.

> * :ref:`bt_mesh_sensor_cli_readme` gets sensor data from one or more :ref:`bt_mesh_sensor_srv_readme`.

Reviewer: ```suggestion
* :ref:`bt_mesh_sensor_cli_readme` gets sensor data from one or more :ref:`Mesh sensor server(s) <bt_mesh_sensor_srv_readme>`.
```

> The model handling is implemented in :file:`src/model_handler.c`. A :c:type:`struct k_delayed_work` item is submitted recursively to periodically request sensor data.

Reviewer: ```suggestion
The model handling is implemented in :file:`src/model_handler.c`.
A :c:type:`struct k_delayed_work` item is submitted recursively to periodically request sensor data.
```

Requirements
************

* One of the following development kits:

  * |nRF52840DK|
  * |nRF52DK|

* Smartphone with Nordic Semiconductor's nRF Mesh mobile app installed in one of the following versions:

  * `nRF Mesh mobile app for Android`_
  * `nRF Mesh mobile app for iOS`_

> * :ref:`bluetooth_mesh_sensor_server` sample application programmed on a separate board and configured according to the Mesh Sensor Server sample's :ref:`bluetooth_mesh_sensor_server_testing` guide.

Reviewer: ```suggestion
* :ref:`bluetooth_mesh_sensor_server` sample application, programmed on a separate board and configured according to the Mesh sensor server sample's :ref:`testing guide <bluetooth_mesh_sensor_server_testing>`
```

User interface
**************

Buttons:
   Can be used to input the OOB authentication value during provisioning.
   All buttons have the same functionality during this procedure.
Terminal:

>    All sensor values gathered from the Server is printed over UART. See :ref:`gs_testing`.

Reviewer: ```suggestion
   All sensor values gathered from the server are printed over UART.
   For more details, see :ref:`gs_testing`.
```

Building and running
********************

.. |sample path| replace:: :file:`samples/bluetooth/mesh/sensor_client`

.. include:: /includes/build_and_run.txt

.. _bluetooth_mesh_sensor_client_testing:

Testing
=======

.. important::
   The Sensor Client sample cannot demonstrate any functionality on its own, and needs a device with the :ref:`bluetooth_mesh_sensor_server` sample running in the same Mesh network.
   Before testing Mesh Sensor Client, go through the Mesh Sensor Server's :ref:`bluetooth_mesh_sensor_server_testing` guide with a different board.

> After programming the sample to your board, you can test the device by provision and configuring it for communication trough the Mesh models.

Author: Done

> .. important::
>    The Sensor Client sample cannot demonstrate any functionality on its own, and needs a device with the :ref:`bluetooth_mesh_sensor_server` sample running in the same Mesh network.
>    Before testing Mesh Sensor Client, go through the Mesh Sensor Server's :ref:`bluetooth_mesh_sensor_server_testing` guide with a different board.

Reviewer: ```suggestion
.. note::
   The Mesh sensor client sample cannot demonstrate any functionality on its own, and needs a device with the :ref:`bluetooth_mesh_sensor_server` sample running in the same mesh network.
   Before testing the Mesh sensor client, go through the Mesh sensor server's :ref:`testing guide <bluetooth_mesh_sensor_server_testing>` with a different board.
```

This can be done using a smartphone with Nordic Semiconductor's nRF Mesh app installed.

> All sensor values gathered from the Server is printed over UART. See :ref:`gs_testing`.

Reviewer: ```suggestion
All sensor values gathered from the server are printed over UART.
For more details, see :ref:`gs_testing`.
```

Provisioning the device
-----------------------

> The provisioning assigns an address range to the device, and adds it to the Mesh network.

Reviewer: ```suggestion
The provisioning assigns an address range to the device, and adds it to the mesh network.
```

Complete the following steps in the nRF Mesh app:

> 1. Tap :guilabel:`Add node` to start scanning for unprovisioned Mesh devices.

Reviewer: ```suggestion
1. Tap :guilabel:`Add node` to start scanning for unprovisioned mesh devices.
```

#. Select the :guilabel:`Mesh Sensor Client` device to connect to it.
#. Tap :guilabel:`Identify` and then :guilabel:`Provision` to provision the device.
#. When prompted, select an OOB method and follow the instructions in the app.

Once the provisioning is complete, the app returns to the Network screen.

Configuring models
------------------

Complete the following steps in the nRF Mesh app to configure models:

1. On the Network screen, tap the :guilabel:`Mesh Sensor Client` node.
   Basic information about the Mesh node and its configuration is displayed.
#. In the Mesh node view, expand the element.
   It contains the list of models in the first and only element of the node.
#. Tap :guilabel:`Sensor Client` to see the model's configuration.
#. Bind the model to application keys to make it open for communication:

   1. Tap :guilabel:`BIND KEY` at the top of the screen.
   #. Select :guilabel:`Application Key 1` from the list.

#. Set the publishing parameters:

   1. Tap :guilabel:`SET PUBLICATION`.
   #. Tap :guilabel:`Publish Address`.
   #. Choose Groups from the dropdown menu.

>    #. Choose an existing group or create a new. *IMPORTANT: The Sensor Server must subscribe to the same grop*.

Author: Landed on keeping using group address after discussion with Trond

>    #. Choose Groups from the dropdown menu.

Reviewer: ```suggestion
   #. Choose Groups from the drop-down menu.
```

>    #. Choose an existing group or create a new. *IMPORTANT: The Sensor Server must subscribe to the same grop*.

Reviewer: ```suggestion
   #. Choose an existing group or create a new.
       *IMPORTANT: The sensor server must subscribe to the same group*.
```
Or you could consider adding a note under this list item instead of writing *IMPORTANT: The sensor server must subscribe to the same group*.

   #. Tap :guilabel:`OK`.
   #. Retransmit Count should be 0 to avoid duplicate logging in the UART terminal.
   #. Press the :guilabel:`check button` at the bottom right corner.

#. Set subscription parameters:

   1. Tap :guilabel:`SUBSCRIBE`.

>    #. Choose an existing group or create a new. *IMPORTANT: The Sensor Server must publish to the same grop*.

Reviewer: ```suggestion
   #. Choose an existing group or create a new.
       *IMPORTANT: The sensor server must publish to the same group*.
```
Same as for the previous comment, you could consider a note.

.. _bluetooth_mesh_sensor_client:

Bluetooth: Mesh sensor client
###############################

> The Bluetooth Mesh sensor client sample demonstrates how to set up a basic Bluetooth Mesh :ref:`bt_mesh_sensor_cli_readme` model application that gets sensor data from one :ref:`bt_mesh_sensor_srv_readme` model.

Author: Did not remove the :ref:`bt_mesh_sensor_cli_readme` link, but changed to :ref:`sensor client <bt_mesh_sensor_cli_readme>`

.. note::
    This sample must be paired with :ref:`bluetooth_mesh_sensor_server` to show any functionality. A client has no sensor data and is dependent on a server to provide it.

Four Mesh sensor types are used in this sample:

1. :cpp:var:`bt_mesh_sensor_present_dev_op_temp` - Published by the server according to its publishing period.
#. :cpp:var:`bt_mesh_sensor_rel_runtime_in_a_dev_op_temp_range` - Periodically requested by the client.
#. :cpp:var:`bt_mesh_sensor_presence_detected` - Published when a button is pressed on the server board.

> #. :cpp:var:`bt_mesh_sensor_time_since_presence_detected` - Periodically requested by the client and published by the server according to its publishing period.

Author: Does not "Periodically requested by the client and published by the server according to its publishing period" qualify as a sentence?

Overview
********

This sample is split into two source files:

* A :file:`main.c` file to handle initialization.

> * One additional file for handling Mesh models, :file:`model_handler.c`.

Reviewer 2: Decision seems to be to have dots.

> * One additional file for handling Mesh models, :file:`model_handler.c`.

Reviewer 3: I reverted them for consistency with other existing samples. Changing this here would require changes on other pages, so I prefer to make the edits in a separate PR, potentially related not only to Mesh.

#. Select the :guilabel:`Mesh Sensor Client` device to connect to it.
#. Tap :guilabel:`Identify` and then :guilabel:`Provision` to provision the device.
#. When prompted, select an OOB method and follow the instructions in the app.

Once the provisioning is complete, the app returns to the Network screen.

Configuring models
------------------

Complete the following steps in the nRF Mesh app to configure models:

1. On the Network screen, tap the :guilabel:`Mesh Sensor Client` node.
   Basic information about the Mesh node and its configuration is displayed.
#. In the Mesh node view, expand the element.
   It contains the list of models in the first and only element of the node.
#. Tap :guilabel:`Sensor Client` to see the model's configuration.
#. Bind the model to application keys to make it open for communication:

   1. Tap :guilabel:`BIND KEY` at the top of the screen.
   #. Select :guilabel:`Application Key 1` from the list.

#. Set the publishing parameters:

   1. Tap :guilabel:`SET PUBLICATION`.
   #. Tap :guilabel:`Publish Address`.
   #. Choose Groups from the drop-down menu.
   #. Choose an existing group or create a new.

>          The sensor server must subscribe to the same grop.

Reviewer: ```suggestion
         The sensor server must subscribe to the same group.
```

   #. Tap :guilabel:`OK`.
   #. Retransmit Count should be 0 to avoid duplicate logging in the UART terminal.
   #. Press the :guilabel:`check button` at the bottom right corner.

#. Set subscription parameters:

   1. Tap :guilabel:`SUBSCRIBE`.
   #. Choose an existing group or create a new.

>          The sensor server must publish to the same grop.

Reviewer: ```suggestion
         The sensor server must publish to the same group.
```

>          The sensor server must publish to the same grop.

Reviewer: ```suggestion
         The sensor server must publish to the same group.
```

.. _bluetooth_mesh_sensor_client:

Bluetooth: Mesh sensor client
#############################

The Bluetooth Mesh sensor client sample demonstrates how to set up a basic Bluetooth Mesh :ref:`sensor client <bt_mesh_sensor_cli_readme>` model application that gets sensor data from one :ref:`sensor server <bt_mesh_sensor_srv_readme>` model.

.. note::
    This sample must be paired with :ref:`bluetooth_mesh_sensor_server` to show any functionality. A client has no sensor data and is dependent on a server to provide it.

Overview
********

This sample is split into the following source files:

* A :file:`main.c` file to handle initialization.
* One additional file for handling Mesh models, :file:`model_handler.c`.

> The following Mesh sensor types are used in this sample:

Reviewer 3: Could you provide a line of explanation after this list or in line 7 (which would be better, I think) about why these various sensor types are used in this sample?

Author: For example "Four different sensor types are used to showcase different ways for the server to publish data. In addition, usage of both single-channel sensors types and series sensor types are shown (reference: http://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/include/bluetooth/mesh/sensor.html?#bt-mesh-sensor-types-channels)."

Reviewer 3: Thanks.

#. When prompted, select an OOB method and follow the instructions in the app.

Once the provisioning is complete, the app returns to the Network screen.

Configuring models
------------------

Complete the following steps in the nRF Mesh app to configure models:

1. On the Network screen, tap the :guilabel:`Mesh Sensor Client` node.
   Basic information about the mesh node and its configuration is displayed.
#. In the Mesh node view, expand the element.
   It contains the list of models in the first and only element of the node.
#. Tap :guilabel:`Sensor Client` to see the model's configuration.
#. Bind the model to application keys to make it open for communication:

   a. Tap :guilabel:`BIND KEY` at the top of the screen.
   #. Select :guilabel:`Application Key 1` from the list.

#. Set the publishing parameters:

   a. Tap :guilabel:`SET PUBLICATION`.
   #. Tap :guilabel:`Publish Address`.
   #. Select :guilabel:`Groups` from the drop-down menu.
   #. Select an existing group or create a new one.

      .. note::
        The sensor server must subscribe to the same group.

   #. Tap :guilabel:`OK`.

>    #. Set the Retransmit Count to zero (:guilabel:`Disabled`) to avoid duplicate logging in the UART terminal.

Reviewer 3: Modified this in both .rst files, but please double check that Disabled matches the app.

Author: Disabled does match the app

Reviewer 3: OK

>    #. Tap :guilabel:`Check` at the bottom right corner.

Reviewer 3: Can you verify this? I had to disinstall the app on my phone, and I'm not sure such Check button exists.

Author: The button is there with a "check icon" (like this: https://en.wikipedia.org/wiki/File:Light_green_check.svg)

Reviewer 3: Roger

#. Set subscription parameters:

   a. Tap :guilabel:`SUBSCRIBE`.
   #. Select an existing group or create a new one.

      .. note::
        The sensor server must publish to the same group.

   #. Tap :guilabel:`OK`.

> #. Tap the back arrow at the top left corner.

Reviewer 3: Is this back arrow a "back" button (thus, we don't need "arrow" and can simply write :guilabel:`Back`) or a custom icon you set?

Also, would be good to have a "to do something" at the end here to give a hint to the reader why tapping this is important.

Author: This is a "back button" to get back to the previous menu. If it is pressed twice you get back to the "main menu". Maybe it is better to write something like this: 
"The sensor client model is now configured and should be able to receive data from the server. Press the back arrow (top left corner) twice to get back to the main screen. "

Reviewer 3: OK
